File: 198-house-robber/house-robber.cpp
class Solution {
public:
    int solve(int ind,vector<int> &nums,vector<int> &dp){
        int n=nums.size();
        if(ind>=n){
            return 0;
        }
        if(dp[ind]!=-1)return dp[ind];
        int take=nums[ind]+solve(ind+2,nums,dp);
        int not_take=solve(ind+1,nums,dp);
        return dp[ind]=max(take,not_take);
    }
    int rob(vector<int>& nums) {
        int n=nums.size();
        vector<int> dp(n,-1);
        return solve(0,nums,dp);
        
        
    }
};

File: 198-house-robber/house-robber.py
class Solution:
    def solve(self,nums,ind,dp):
        if(ind>=len(nums)):
            return 0
        if(dp[ind]!=-1):
            return dp[ind]
        take=nums[ind]+self.solve(nums,ind+2,dp)
        not_take=self.solve(nums,ind+1,dp)
        dp[ind]=max(take,not_take)
        return dp[ind]

    def rob(self, nums: List[int]) -> int:
        dp=[-1]*(len(nums)+1)
        res=self.solve(nums,0,dp)
        return res
        